distributed/shuffle/_rechunk.py

```diff
@@ -216,27 +228,215 @@ def rechunk_p2p(x: da.Array, chunks: ChunkedAxes) -> da.Array:
+
+        self.name = name
+        self.token = token
+        self.chunks = chunks
+        self.chunks_input = chunks_input
+        self.name_input = name_input
+        self.disk = disk
+        if keepmap is not None:
+            self.keepmap = keepmap
+        else:
+            shape = tuple(len(axis) for axis in chunks)
+            self.keepmap = np.ones(shape, dtype=bool)
+        self._cached_dict = None
+        super().__init__(annotations=annotations)
+
+    def __repr__(self) -> str:
+        return f"{type(self).__name__}<name='{self.name}', chunks={self.chunks}>"
+
+    def get_output_keys(self) -> set[Key]:
+        import numpy as np
+
+        return {
+            (self.name,) + nindex
+            for nindex in np.ndindex(tuple(len(axis) for axis in self.chunks))
+        }
+
+    def is_materialized(self) -> bool:
+        return self._cached_dict is not None
+
+    @property
+    def _dict(self) -> _T_LowLevelGraph:
+        """Materialize full dict representation"""
+        dsk: _T_LowLevelGraph
+        if self._cached_dict is not None:
+            return self._cached_dict
+        else:
+            dsk = self._construct_graph()
+            self._cached_dict = dsk
+        return self._cached_dict
+
+    def __getitem__(self, key: Key) -> tuple:
+        return self._dict[key]
+
+    def __iter__(self) -> Iterator[Key]:
+        return iter(self._dict)
+
+    def __len__(self) -> int:
+        return len(self._dict)
+
+    def _cull(self, keepmap: np.ndarray) -> P2PRechunkLayer:
+        return P2PRechunkLayer(
+            name=self.name,
+            token=self.token,
+            chunks=self.chunks,
+            chunks_input=self.chunks_input,
+            name_input=self.name_input,
+            disk=self.disk,
+            keepmap=keepmap,
+            annotations=self.annotations,
+        )
+
+    def _keys_to_indices(self, keys: Iterable[Key]) -> set[tuple[int, ...]]:
+        """Simple utility to convert keys to chunk indices."""
+        chunks = set()
+        for key in keys:
+            if not isinstance(key, tuple) or len(key) < 2 or key[0] != self.name:
+                continue
+            chunk = cast(tuple[int, ...], key[1:])
+            assert all(isinstance(index, int) for index in chunk)
+            chunks.add(chunk)
+        return chunks
+
+    def cull(
+        self, keys: set[Key], all_keys: Collection[Key]
+    ) -> tuple[P2PRechunkLayer, dict]:
+        """Cull a P2PRechunkLayer HighLevelGraph layer.
+
+        The underlying graph will only include the necessary
+        tasks to produce the keys (indices) included in `keepmap`.
+        Therefore, "culling" the layer only requires us to reset this
+        parameter.
+        """
+        import numpy as np
+
+        keepmap = np.zeros_like(self.keepmap)
+        indices_to_keep = self._keys_to_indices(keys)
+
+        from dask.array.rechunk import old_to_new
+
+        _old_to_new = old_to_new(self.chunks_input, self.chunks)
```

Comment:
Cache this?

Reply:
I'll run some tests to figure out if this is bad, from what I understand, we should only cull once, i.e., this would get called twice in total.